ui/appui-react/src/appui-react/messages/ActivityMessage.tsx

```diff
@@ -26,37 +34,110 @@ export interface ActivityMessageProps {
+            </MessageButton>
+          </div>
+        ) : (
+          <MessageButton onClick={props.dismissActivityMessage}>
+            <Icon iconSpec="icon-close" />
+          </MessageButton>
+        )}
+        progress={(messageDetails && messageDetails.showProgressBar) &&
             <MessageProgress
               status={Status.Information}
               progress={props.activityMessageInfo.percentage}
             />
+        }
+      >
+        <div>
+          <MessageLabel message={props.activityMessageInfo.message} className="uifw-statusbar-message-brief" />
+          {(messageDetails && messageDetails.showPercentInMessage) &&
+            <Small>{props.activityMessageInfo.percentage + percentCompleteLabel}</Small>
           }
-        >
-          <div>
-            <MessageLabel message={props.activityMessageInfo.message} className="uifw-statusbar-message-brief" />
-            {(messageDetails && messageDetails.showPercentInMessage) &&
-              <Small>{props.activityMessageInfo.percentage + percentCompleteLabel}</Small>
-            }
-          </div>
-        </MessageLayout>
+        </div>
+      </MessageLayout>
+    </Message>
+  );
+}
+
+interface CustomActivityMessageProps {
+  cancelActivityMessage?: () => void;
+  dismissActivityMessage?: () => void;
+  activityMessageInfo?: ActivityMessageEventArgs;
+  settings?: ToasterSettings;
+}
+
+/**
+ * Hook to render an Activity message.
+ */
+export function useActivityMessage({activityMessageInfo, dismissActivityMessage, cancelActivityMessage, settings}: CustomActivityMessageProps) {
+  const [cancelLabel] = React.useState(UiCore.translate("dialog.cancel"));
+  const recentToast = React.useRef<{close: () => void} | undefined>();
+  React.useEffect(() => {
+    toaster.setSettings(settings ?? {placement: "top"});
+  }, [settings]);
+
+  React.useEffect(() => {
+    if (activityMessageInfo?.restored) {
+      recentToast.current = toaster.informational(
+        <CustomActivityMessageContent initialActivityMessageInfo={activityMessageInfo} />,
+        { onRemove: dismissActivityMessage, type: "persisting", link: activityMessageInfo?.details?.supportsCancellation && cancelActivityMessage ? { title: cancelLabel, onClick: cancelActivityMessage } : undefined}
+      );
+    }
+    if (!activityMessageInfo) {
+      recentToast.current?.close();
+    }
+  }, [activityMessageInfo, cancelActivityMessage, cancelLabel, dismissActivityMessage]);
+}
+
+/**
+ * Component wrapping the `useActivityMessage` hook to use in class components.
+ */
+export function CustomActivityMessageRenderer({activityMessageInfo, dismissActivityMessage, cancelActivityMessage, settings}: CustomActivityMessageProps) {
+  useActivityMessage({activityMessageInfo, cancelActivityMessage, dismissActivityMessage, settings});
+
+  return <></>;
+}
+
+function CustomActivityMessageContent({initialActivityMessageInfo}: {initialActivityMessageInfo?: ActivityMessageEventArgs})  {
```

Comment:
This is custom component for UI of ActivityMessage. I needed this to be passed for `toaster`.
It also subsribes to events just to rerender the % part. There is no way to say `toaster` to rerender existing Toast.